galaxy_ng/app/api/v3/viewsets/collection.py

```diff
@@ -87,21 +92,66 @@ class CollectionViewSet(api_base.LocalSettingsMixin,
     serializer_class = CollectionSerializer
 
 
-class UnpaginatedCollectionVersionViewSet(api_base.LocalSettingsMixin,
-                                          ViewNamespaceSerializerContextMixin,
-                                          pulp_ansible_views.UnpaginatedCollectionVersionViewSet):
+class UnpaginatedCollectionVersionViewSet(
+    api_base.LocalSettingsMixin,
+    ViewNamespaceSerializerContextMixin,
+    DestroyModelMixin,
+    pulp_ansible_views.UnpaginatedCollectionVersionViewSet,
+):
     pagination_class = None
     serializer_class = UnpaginatedCollectionVersionSerializer
     permission_classes = [access_policy.CollectionAccessPolicy]
 
 
+def get_dependents(parent):
+    """Given a parent collection, return a list of collections that depend on it."""
+    key = f"{parent.namespace}.{parent.name}"
+    dependents = []
+    for child in CollectionVersion.objects.filter(dependencies__has_key=key):
+        spec = SpecifierSet(child.dependencies[key])
+        if Version(parent.version) in spec:
+            dependents.append(child)
+    return dependents
+
+
 class CollectionVersionViewSet(api_base.LocalSettingsMixin,
                                ViewNamespaceSerializerContextMixin,
                                pulp_ansible_views.CollectionVersionViewSet):
     serializer_class = CollectionVersionSerializer
     permission_classes = [access_policy.CollectionAccessPolicy]
     list_serializer_class = CollectionVersionListSerializer
 
+    def destroy(self, request: Request, *args, **kwargs) -> Response:
+        """
+        Allow a CollectionVersion to be deleted.
+        1. Perform Dependency Check to verify that the collection version can be deleted
+        2. If it can, perform Collection Repository Cleanup
+        3. If the collection version can’t be deleted, return the reason why
+        4. If the version being deleted is the last collection version in the collection,
+           remove the collection object as well.
+        """
+        collection_version = self.get_object()
+        # dependency check
+        dependents = get_dependents(collection_version)
+        if dependents:
+            return Response(
+                {
+                    "detail": _(
+                        "Collection {namespace}.{name} could not be deleted "
+                        "because there are other collections that require it."
+                    ).format(
+                        namespace=collection_version.namespace,
+                        name=collection_version.collection.name,
+                    ),
+                    "dependent_collections": [f"{dep.namespace}.{dep.name}" for dep in dependents],
+                },
+                status=400,
+            )
+        # collection repo cleanup
+
```

Comment:
@awcrosby the **Collection Repo Cleanup** tasks are described in Content Deletion arch doc.

on this under review PR https://github.com/ansible/galaxy_ng/pull/958 there is a script to generate some collections with dependencies.

```bash
django-admin create-test-collections --strategy=test --prefix=710 -y
```

You gonna feed your system with collections and dependencies to test.